crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -127,83 +125,101 @@ pub struct ModuleResolutionSettings {
+    /// Each `.unwrap()` call is a point where we're validating a setting that the user would pass
+    /// and transforming it into an internal representation for a validated path.
+    /// Rather than panicking if a path fails to validate, we should display an error message to the user
+    /// and exit the process with a nonzero exit code.
+    /// This validation should probably be done outside of Salsa?
+    fn into_resolved_settings(self) -> ResolvedModuleResolutionSettings {
         let ModuleResolutionSettings {
+            target_version,
             extra_paths,
             workspace_root,
             site_packages,
             custom_typeshed,
         } = self;
 
-        let mut paths: Vec<_> = extra_paths
+        let mut paths: Vec<ModuleResolutionPathBuf> = extra_paths
             .into_iter()
-            .map(|path| ModuleSearchPath::new(path, ModuleSearchPathKind::Extra))
+            .map(|fs_path| ModuleResolutionPathBuf::extra(fs_path).unwrap())
             .collect();
 
-        paths.push(ModuleSearchPath::new(
-            workspace_root,
-            ModuleSearchPathKind::FirstParty,
-        ));
+        paths.push(ModuleResolutionPathBuf::first_party(workspace_root).unwrap());
 
-        // TODO fallback to vendored typeshed stubs if no custom typeshed directory is provided by the user
         if let Some(custom_typeshed) = custom_typeshed {
-            paths.push(ModuleSearchPath::new(
-                custom_typeshed.join(TYPESHED_STDLIB_DIRECTORY),
-                ModuleSearchPathKind::StandardLibrary,
-            ));
+            paths.push(
+                ModuleResolutionPathBuf::stdlib_from_typeshed_root(&custom_typeshed).unwrap(),
+            );
         }
 
         // TODO vendor typeshed's third-party stubs as well as the stdlib and fallback to them as a final step
         if let Some(site_packages) = site_packages {
-            paths.push(ModuleSearchPath::new(
-                site_packages,
-                ModuleSearchPathKind::SitePackagesThirdParty,
-            ));
+            paths.push(ModuleResolutionPathBuf::site_packages(site_packages).unwrap());
         }
 
-        OrderedSearchPaths(paths)
+        ResolvedModuleResolutionSettings {
+            target_version,
+            search_paths: OrderedSearchPaths(paths.into_iter().map(Arc::new).collect()),
+        }
     }
 }
 
 /// A resolved module resolution order, implementing PEP 561
 /// (with some small, deliberate differences)
 #[derive(Clone, Debug, Default, Eq, PartialEq)]
-pub(crate) struct OrderedSearchPaths(Vec<ModuleSearchPath>);
+pub(crate) struct OrderedSearchPaths(Vec<Arc<ModuleResolutionPathBuf>>);
 
 impl Deref for OrderedSearchPaths {
-    type Target = [ModuleSearchPath];
+    type Target = [Arc<ModuleResolutionPathBuf>];
 
     fn deref(&self) -> &Self::Target {
         &self.0
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct ResolvedModuleResolutionSettings {
```

Comment:
`ResolvedModuleResolutionSettings` is a kind of unfortunate name (due to the double use of "Resolved/Resolution" for very different meanings within the same name).

Could this just be `ModuleResolutionSettings` and the raw version be `RawModuleResolutionSettings` or `ModuleResolutionSettingsInput`?

Reply:
I would leave it as is. Input is misleading because it gives the impression that it is a salsa input where it is not. We discussed alternatives, for example using a builder that directly constructs the "resolved" thing but decided to leave this for another day. I suspect that all of this will change anyway when we introduce settings